worldedit-bukkit/src/main/java/com/fastasyncworldedit/bukkit/adapter/NMSAdapter.java

```diff
@@ -140,4 +144,122 @@ public void sendChunk(IChunkGet chunk, int mask, boolean lighting) {
         ((BukkitGetBlocks) chunk).send();
     }
 
+    private static final ConcurrentHashMap<String, ConcurrentHashMap<IntPair, ChunkLock>> SENDING_CHUNKS = new ConcurrentHashMap<>();
+
+    /**
+     * Atomically set the given chunk section to the chunk section array stored in the chunk, given the expected existing chunk
+     * section instance at the given layer position.
+     * <p>
+     * Acquires a (FAWE-implemented only) write-lock on the chunk packet lock, waiting if required before writing, then freeing
+     * the lock. Also sets a boolean to indicate a write is waiting and therefore reads should not occur.
+     * <p>
+     * Utilises ConcurrentHashMap#compute for easy synchronisation for all of the above. Only tryWriteLock is used in blocks
+     * synchronised using ConcurrentHashMap methods.
+     *
+     * @since TODO
+     */
+    protected static <LevelChunkSection> boolean setSectionAtomic(
+            String worldName,
+            IntPair pair,
+            LevelChunkSection[] sections,
+            LevelChunkSection expected,
+            LevelChunkSection value,
+            int layer
+    ) {
+        if (layer < 0 || layer >= sections.length) {
+            return false;
+        }
+        long[] stamp = new long[1];
+        ChunkLock[] chunkLock = new ChunkLock[1];
+        ConcurrentHashMap<IntPair, ChunkLock> chunks = SENDING_CHUNKS.computeIfAbsent(
+                worldName,
+                world -> new ConcurrentHashMap<>()
+        );
+        chunks.compute(pair, (k, lock) -> {
+            if (lock == null) {
+                lock = new ChunkLock();
+            } else if (lock.writeWaiting) {
+                throw new IllegalStateException("Attempting to write chunk section when write is already ongoing?!");
+            }
+            stamp[0] = lock.lock.tryWriteLock();
+            chunkLock[0] = lock;
+            lock.writeWaiting = true;
+            return lock;
+        });
+        try {
+            if (stamp[0] == 0) {
+                stamp[0] = chunkLock[0].lock.writeLock();
+            }
+            return ReflectionUtils.compareAndSet(sections, expected, value, layer);
+        } finally {
+            chunks = SENDING_CHUNKS.get(worldName);
```

Comment:
is there any scenario where this is actually needed? The old reference should still be the correct one as we never remove it. Alternatively, did you consider using a `ConcurrentHashMap<*WorldIntPair*, ChunkLock>` instead? This way we won't have unused CHM instances laying around (which could probably be a problem with many temporary worlds).

Reply:
The idea was to reduce lock contention between worlds when locking/unlocking. I suppose the operations themselves shouldn't take too long, though perhaps it's worth having the CHM in a BukkitWorld instance so it's GCed as needed if still attempting to do this

Reply:
Moving it into BukkitWorld sounds good.

Reply:
FAWE does not actually cache/limit the number of BukkitWorld instances, and given upstream does not either, it feels quite wrong to do this. I don't know if it's worth wrapping a WeakHashMap into a synchronised map and using the NMS world (or a normal bukkit world) as a key?